### ISSUE TITLE
PIZ Compression

### ISSUE THREAD

**Issue by johannesvollmer** [johannesvollmer/exrs#44]
This is the tracking issue for adding PIZ compression.

# Current State
Checkout the [PIZ branch](https://github.com/johannesvollmer/exrs/tree/piz/src/compression/piz):
- Decoding: 100% 
- Encoding: 100%

References:
- [openexr compressor](https://github.com/AcademySoftwareFoundation/openexr/blob/master/OpenEXR/IlmImf/ImfPizCompressor.cpp)
- [openexr huffman decoding](https://github.com/AcademySoftwareFoundation/openexr/blob/88246d991e0318c043e6f584f7493da08a31f9f8/OpenEXR/IlmImf/ImfHuf.cpp)
- [openexr wavelet decoding](https://github.com/AcademySoftwareFoundation/openexr/blob/8cd1b9210855fa4f6923c1b94df8a86166be19b1/OpenEXR/IlmImf/ImfWav.cpp)

**Comment by johannesvollmer** [comment]
I use this method to run a single test for the PIZ compression method:

https://github.com/johannesvollmer/exrs/blob/0f4910953552a397b658bcdc3b98b4f81687c806/tests/dev.rs#L49-L56

@dgsantana You can use the command `test --package exr --test dev test_roundtrip -- --exact --ignored`  to run this method if you find this useful.

**Comment by dgsantana** [comment]
Ok thanks, that is very helpful, the first time I ran the test, it was an
endless wall of debug info :).

Johannes Vollmer <notifications@github.com> escreveu no dia quinta,
14/05/2020 à(s) 17:17:

**Comment by johannesvollmer** [comment]
Yes, sorry, I did not expect anyone actually having a look at the piz branch, haha! It's some desparate logging that I did in order to try and understand why it does not work, but that logging didn't help me all that much

**Comment by dgsantana** [comment]
My idea, after looking at the code, is to try and build in "blocks" and comparing with the C++ output (ground truth). To see where are the differences are coming, it will be a tedious job, but since there is a lot of pointer math on C++, I think it maybe something there.

**Comment by johannesvollmer** [comment]
Yes, that would be a reasonable approach. I guess I can't avoid compiling the original OpenEXR library on my computer any longer

**Comment by dgsantana** [comment]
But I think you did an amazing job. Just the fact that as-is it works and doesn't depend on any binding, I find it amazing.

**Comment by johannesvollmer** [comment]
Thank you! I appreciate it.

**Comment by dgsantana** [comment]
> 
> 
> Yes, that would be a reasonable approach. I guess I can't avoid compiling the original OpenEXR library on my computer any longer

I will try to avoid that :D, although I have it on my machine due to other projects I daily build (Blender, is one of them). I will try to build simple C++ executables that contain the code (bare minimal) to compare results and use those results for tests on Rust, this should avoid having too many moving parts. I think the Huffman codec can be easily "detached" from the openexr code, to create such an test executable.

**Comment by johannesvollmer** [comment]
Great idea! The actual `PIZ` part might already be working, at least the "roundtrip" tests for that already works. Having huffman working would bring us like 90% to a working product I think

**Comment by dgsantana** [comment]
Just to let you know I got huffman compression working. Will work on decompression in the next few days.

**Comment by johannesvollmer** [comment]
Awesome :)

**Comment by dgsantana** [comment]
Also added a ground-truth test (not yet pushed to my fork) for the compression, based on 100 bytes huffman compression test of the official OpenEXR.

**Comment by johannesvollmer** [comment]
Excellent, I think I'm going to do that for the wavelet transform, if I can find some tests

**Comment by johannesvollmer** [comment]
I just ruthlessly pushed some changes to the PIZ branch, I hope that's ok, I only changed the `wavelet.rs` module, adding some ground-truth tests from the OpenEXR

**Comment by dgsantana** [comment]
Don't worry, I working only on the huffman.rs for now.

**Comment by johannesvollmer** [comment]
Hey, considering that you got Huffman working, what else has to be done until PIZ works? If there is something else not working, I'd love to give it a try too, maybe we can already merge the huffman.rs file into master?

**Comment by dgsantana** [comment]
I haven't finished the decompression part. I can merge what I have into the piz branch.

**Comment by johannesvollmer** [comment]
Oh, I misunderstood you earlier, I thought both directions were already working. Sorry, I didn't mean to stress you out, haha :D

There is no need to merge until decompression works as well, don't worry :)

**Comment by dgsantana** [comment]
> 
> 
> Oh, I misunderstood you earlier, I thought both directions were already working. Sorry, I didn't mean to stress you out, haha :D
> 
> There is no need to merge until decompression works as well, don't worry :)

Don't worry, it's been a crazy weekend. I'm also taking some days off company work, so it may take a little longer. But I wanted to included this compression as an option on our tool that depends on exrs.

**Comment by johannesvollmer** [comment]
Yes, of course. Works for me!

**Comment by johannesvollmer** [comment]
While implementing the PIZ functions that will later call the huffman and wavelet module, I realized it would be nice if the `huffman::compress` function signature was a little different, like this:

```rust
pub fn compress(uncompressed: &[u16]) -> Result<Vec<u8>>
```

Can we agree on that?

**Comment by dgsantana** [comment]
> 
> 
> While implementing the PIZ functions that will later call the huffman and wavelet module, I realized it would be nice if the `huffman::compress` function signature was a little different, like this:
> 
> ```rust
> pub fn compress(uncompressed: &[u16]) -> Result<Vec<u8>>
> ```
> 
> Can we agree on that?

Agreed, and easy to change.

**Comment by dgsantana** [comment]
I had compression and decompression working, but found a rare bug on long code in the Huffman decompression. I trying to find out what it's causing, it's the only thing holding the pull request. Also, some parts of the code, are not yet very rusty (too much 1-1 to C++), and can probably be refactor later on.

**Comment by johannesvollmer** [comment]
Ok nice, refactoring is definitely easier once that is already working

**Comment by dgsantana** [comment]
Found the bug. It's on the FrequencyHeap, on the compression part, in some cases the MinMaxHeap that I was using doesn't produce the same sorted result, which is weird. I'm looking into another implementation, since make_heap in C++ as some weird side effects, not resorting push elements in the current heap, it's like pushes all resort new elements, not changing the old ones. I used a hack to simulate this.

**Comment by johannesvollmer** [comment]
Yes, the underlying heap uses a plain `std::vec` under the hood, which makes things really ... specific I guess. Have you tried using a plain `std::collections::BinaryHeap`? Don't know if that would be appropriate, just asking

**Comment by dgsantana** [comment]
Yup, I tried, but since they are using a min-heap, and we can't use a comparator function, to replicate it requires to push all values to the heap inside another struct with ordering, and it gets weird. But I think I going to give it another go.

**Comment by johannesvollmer** [comment]
Sounds really annoying, good luck :D

**Comment by dgsantana** [comment]
Finally fixed, used the BinaryHeap with a custom struct and a custom comparator. Tests are passing, and byte by byte comparison with official OpenEXR is good.

**Comment by johannesvollmer** [comment]
Nice I'm going to have a look now

**Comment by johannesvollmer** [comment]
I think writing images works: If you run the example `generate_rgba` in the branch (which in this branch uses PIZ) the image can be opened in affinity photo

**Comment by johannesvollmer** [comment]
However reading an image does not work yet, probably an issue with in the actual `piz/mod.rs`. I'm investigating...

**Comment by johannesvollmer** [comment]
Tests indicate that the piz module itself actually works for f16 but not f32, but whole images never work

**Comment by johannesvollmer** [comment]
Ok, I got roundtrip tests for piz.rs working also for f32, but whole images still do not work yet

**Comment by johannesvollmer** [comment]
Ok so:

The piz module roundtrip tests pass. The huffman roundtrip tests and ground-truth tests pass. The wavelet round trip and ground-truth tests also pass. Writing PIZ images works, apparently.

But for whole files, the roundtrip tests in `tests/dev.rs` fails. This indicates that the error must be while reading whole files, probably even in `compression/mod.rs` or `meta/mod.rs::Header`. 

But I just checked that all non-piz files still pass, meaning the error must be piz specific. The error message when testing roundtrip is from huffman: `decoded data are longer than expected`. But the roundtrip tests work, so this error is probably in the code that prepares the data for huffman. But the piz module roundtrip tests work, so the piz module probably is not the problem. Aaaaaargh :D

**Comment by dgsantana** [comment]
:D. I will have a look later, today I'm taking care of the jungle on my garden, and since I have some time off from my company work, I will give it a try and see if I can find the cause.

**Comment by johannesvollmer** [comment]
No need to hurry :) Actually I'm currently writing a bachelor's thesis, and I will also not have too much time to spend on this the next the weeks. After that, I'm all in

**Comment by johannesvollmer** [comment]
When trying to read all piz files in the repository:

```
Err(Invalid("compressed data (PIZ)")): "tests/images/valid\\custom\\crowskull\\crow_piz.exr"
Err(Invalid("compressed data (PIZ)")): "tests/images/valid\\custom\\crowskull\\crow_piz_noisy_rgb.exr"
Err(Invalid("compressed data (PIZ)")): "tests/images/valid\\openexr\\Chromaticities\\Rec709.exr"
Err(Invalid("compressed data (PIZ)")): "tests/images/valid\\openexr\\MultiResolution\\WavyLinesSphere.exr"
Err(Invalid("compressed data (PIZ)")): "tests/images/valid\\openexr\\ScanLines\\CandleGlass.exr"
Err(Invalid("compressed data (PIZ)")): "tests/images/valid\\openexr\\ScanLines\\Desk.exr"
Err(Invalid("compressed data (PIZ)")): "tests/images/valid\\openexr\\ScanLines\\MtTamWest.exr"
Err(Invalid("compressed data (PIZ)")): "tests/images/valid\\openexr\\ScanLines\\PrismsLenses.exr"
Err(Invalid("compressed data (PIZ)")): "tests/images/valid\\openexr\\ScanLines\\StillLife.exr"
Err(Invalid("compressed data (PIZ)")): "tests/images/valid\\openexr\\ScanLines\\Tree.exr"
Err(Invalid("compressed data (PIZ)")): "tests/images/valid\\openexr\\DisplayWindow\\t01.exr"
Err(Invalid("compressed data (PIZ)")): "tests/images/valid\\openexr\\DisplayWindow\\t02.exr"
Err(Invalid("compressed data (PIZ)")): "tests/images/valid\\openexr\\DisplayWindow\\t03.exr"
Err(Invalid("compressed data (PIZ)")): "tests/images/valid\\openexr\\DisplayWindow\\t04.exr"
Err(Invalid("compressed data (PIZ)")): "tests/images/valid\\openexr\\DisplayWindow\\t05.exr"
Err(Invalid("compressed data (PIZ)")): "tests/images/valid\\openexr\\DisplayWindow\\t06.exr"
Err(Invalid("compressed data (PIZ)")): "tests/images/valid\\openexr\\DisplayWindow\\t07.exr"
Err(Invalid("compressed data (PIZ)")): "tests/images/valid\\openexr\\DisplayWindow\\t08.exr"
Err(Invalid("compressed data (PIZ)")): "tests/images/valid\\openexr\\DisplayWindow\\t09.exr"
Err(Invalid("compressed data (PIZ)")): "tests/images/valid\\openexr\\DisplayWindow\\t10.exr"
Err(Invalid("compressed data (PIZ)")): "tests/images/valid\\openexr\\DisplayWindow\\t11.exr"
Err(Invalid("compressed data (PIZ)")): "tests/images/valid\\openexr\\DisplayWindow\\t12.exr"
Err(Invalid("compressed data (PIZ)")): "tests/images/valid\\openexr\\DisplayWindow\\t13.exr"
Err(Invalid("compressed data (PIZ)")): "tests/images/valid\\openexr\\DisplayWindow\\t14.exr"
Err(Invalid("compressed data (PIZ)")): "tests/images/valid\\openexr\\DisplayWindow\\t15.exr"
Err(Invalid("compressed data (PIZ)")): "tests/images/valid\\openexr\\DisplayWindow\\t16.exr"
Ok("Success!"): "tests/images/valid\\openexr\\IlmfmlmflmTest\\comp_b44_piz.exr"
Ok("Success!"): "tests/images/valid\\openexr\\IlmfmlmflmTest\\lineOrder_decreasing.exr"
Ok("Success!"): "tests/images/valid\\openexr\\IlmfmlmflmTest\\lineOrder_increasing.exr"
Ok("Success!"): "tests/images/valid\\openexr\\IlmfmlmflmTest\\test_native1.exr"
Ok("Success!"): "tests/images/valid\\openexr\\IlmfmlmflmTest\\test_native2.exr"
Ok("Success!"): "tests/images/valid\\openexr\\LuminanceChroma\\Garden.exr"
Ok("Success!"): "tests/images/valid\\openexr\\MultiView\\Balls.exr"
Ok("Success!"): "tests/images/valid\\openexr\\MultiView\\Fog.exr"
Ok("Success!"): "tests/images/valid\\openexr\\MultiView\\LosPadres.exr"
Ok("Success!"): "tests/images/valid\\openexr\\TestImages\\AllHalfValues.exr"
Ok("Success!"): "tests/images/valid\\openexr\\Tiles\\GoldenGate.exr"
Err(Invalid("compressed data (PIZ)")): "tests/images/valid\\openexr\\IlmfmlmflmTest\\comp_dwaa_piz.exr"
Err(Invalid("compressed data (PIZ)")): "tests/images/valid\\openexr\\IlmfmlmflmTest\\comp_dwab_piz.exr"
Err(Invalid("compressed data (PIZ)")): "tests/images/valid\\openexr\\IlmfmlmflmTest\\comp_piz.exr"
Err(NotSupported("channel subsampling not supported yet")): "tests/images/valid\\openexr\\LuminanceChroma\\MtTamNorth.exr"
Err(NotSupported("channel subsampling not supported yet")): "tests/images/valid\\openexr\\LuminanceChroma\\StarField.exr"
Err(NotSupported("channel subsampling not supported yet")): "tests/images/valid\\openexr\\Chromaticities\\Rec709_YC.exr"
Ok("Success!"): "tests/images/valid\\openexr\\Chromaticities\\XYZ.exr"
Err(NotSupported("channel subsampling not supported yet")): "tests/images/valid\\openexr\\Chromaticities\\XYZ_YC.exr"
```
Some are decoded successfully. What is the difference between those that work and those that don't?

**Comment by johannesvollmer** [comment]
So, I inserted roundtrip tests into the actual decompression routine, and it fails. When writing a file, decompressing the just compressed data will not produce the original data.

```rust
    let huffman_compressed: Vec<u8> = huffman::compress(&tmp)?;

    #[cfg(debug_assertions)] {
        let mut decompressed = Vec::new();
        huffman::decompress(&huffman_compressed, &mut decompressed).unwrap();
        assert_eq!(decompressed, tmp)
    }
```

I'll extract that data and insert it into new huffman tests

**Comment by johannesvollmer** [comment]
The piz branch should now contain a huffman.rs with an additional test, which has data that cannot be compressed and then decompressed

**Comment by johannesvollmer** [comment]
I was able to reduce it down to the sequence of zeroes being problematic. Compressing an array of 11 zeroes works, but 12 zeroes fail.

**Comment by dgsantana** [comment]
> I was able to reduce it down to the sequence of zeroes being problematic. Compressing an array of 11 zeroes works, but 12 zeroes fail.

That is really weird. I will have a look, it must be something on the build huffman coding, could be one more problem on the min-heap.

**Comment by johannesvollmer** [comment]
Awesome :)

**Comment by johannesvollmer** [comment]
Some of the existing exr images with piz compression cannot be read, indicating a problem with the decoding algorithm, maybe not even the encoding

**Comment by dgsantana** [comment]
I have been a bit busy, but I will try to have a go. I still haven't checked the problem with the 12 zeros.

**Comment by johannesvollmer** [comment]
Jo Problem, Just leaving some thoughts and notes here for later :) don't worry

**Comment by dgsantana** [comment]
Found the issue, it's a simple cast, from i64 to u8 :).

**Comment by dgsantana** [comment]
False alarm, there is more to it than just this. The decoding table seems to have a different order.
In the sample code I'm using for both the rust code and c++ version, getting the index of the shortcode yields a different result.
C++
```
Adding short code 0 with len 1 at [8192..16384]
Adding short code 65534 with len 2 at [4096..8192]
Adding short code 65535 with len 3 at [0..2048]
Adding short code 65536 with len 3 at [2048..4096]

Reading ShortCode I:2155 V:65536 L:3 R:65536 C:37078 CL:12
```

Rust
```
Adding short code 0 with len 1 at [8192..16384]
Adding short code 65534 with len 2 at [4096..8192]
Adding short code 65535 with len 3 at [0..2048]
Adding short code 65536 with len 3 at [2048..4096]

Reading ShortCode I:2155 V:0 L:3 R:65536 C:37078 CL:12
```

The decoding "seems" equal, but for some reason index 2155 which should give 65536, returns 0.
:(

**Comment by dgsantana** [comment]
Ok found the bug, it seems that castung 65535usize to u16 causes it to be 0. This shouldn't happen, since 65535 is the upper limit of u16.

**Comment by dgsantana** [comment]
Ok, the problem is that code which is stored in a struct in c++ with 24 bits (int), while we are using u16. Refactoring code.

**Comment by dgsantana** [comment]
I tried all images on the dev test_roundtrip, and they are passing.

```
file tests/images/valid/custom/crowskull/crow_piz_noisy_rgb.exr
starting read 1... ...read 1 successfull
starting write... ...write successfull: 12.44825mb
starting read 2... ...read 2 successfull

file tests/images/valid/openexr/TestImages/GammaChart.exr
starting read 1... ...read 1 successfull
starting write... ...write successfull: 0.080591mb
starting read 2... ...read 2 successfull

file tests/images/valid/openexr/Tiles/GoldenGate.exr
starting read 1... ...read 1 successfull
starting write... ...write successfull: 3.693668mb
starting read 2... ...read 2 successfull

file tests/images/valid/custom/crowskull/crow_pxr24.exr
starting read 1... ...read 1 successfull
starting write... ...write successfull: 3.372144mb
starting read 2... ...read 2 successfull

file tests/images/valid/custom/crowskull/crow_rle.exr
starting read 1... ...read 1 successfull
starting write... ...write successfull: 3.37022mb
starting read 2... ...read 2 successfull

file tests/images/valid/custom/crowskull/crow_zip_half.exr
starting read 1... ...read 1 successfull
starting write... ...write successfull: 1.860244mb
starting read 2... ...read 2 successfull

file tests/images/valid/custom/crowskull/crow_piz.exr
starting read 1... ...read 1 successfull
starting write... ...write successfull: 3.37022mb
starting read 2... ...read 2 successfull
test test_roundtrip ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 3 filtered out
```

**Comment by dgsantana** [comment]
Also as a note, compressed files can be bigger then uncompressed (I noted that you have a debug_assert to check the compressed < uncompressed / 2). Huffman structures, can add to the payload, e.g.: 100 bytes uncompressed, result in 703 compressed bytes.

**Comment by johannesvollmer** [comment]
Awesome, thanks!

The size assertion should theoretically hold true when reading a file (where the compressed data is larger than uncompressed, the exr file will simply contain the uncompressed data instead). But that's not really important, we should remove it.

I don't quite want to publish yet though - I'll do some fuzzy testing and we should do minor simple refactorings before going live. This includes making code more rusty, adding more assertions in general, making sure that error handling works, and so on. Frankly, I'm in the middle of something else right now, but I'll come back to you in a few days. If you already want to do some refactorings, I won't stop you. Anyways, thanks for the great work so far! :)

**Comment by dgsantana** [comment]
Agreed to make it rustier 😄.

I did notice a "bug" running the fuzz tests. Since I move all types `i64` (`Int64` in C++ as I said in the commit is unsigned) to `u64`, there is a special case `code_bit_count`, in C++ code it's an `int`. This makes negative values possible, changing this in our case solves the error in the fuzz tests, but I'm not confident that the `code_bit_count` should have negative values, but again, I'm no expert in Huffman encoding 😄.

**Comment by johannesvollmer** [comment]
> `Int64` in C++ as I said in the commit is unsigned) 

Can you please show me where you found this? Everywhere else in this library, the code assumes that C++ `int` is an `i32`, could this also be wrong?

**Comment by dgsantana** [comment]
Here `include\OpenEXR\ImathInt64.h`

```
#if (defined _WIN32 || defined _WIN64) && _MSC_VER >= 1300
    typedef unsigned __int64 Int64;
    typedef __int64 SInt64;
#elif ULONG_MAX == 18446744073709551615LU
    typedef long unsigned int Int64;
    typedef long int SInt64;
#else
    typedef long long unsigned int Int64;
    typedef long long int SInt64;
#endif
```

**Comment by dgsantana** [comment]
But `int` is indeed the normal C++ int32 so `i32` is correct. We did have a case where there we used u16 in the Code, and in C++ it's defined as an int but with only 24bits.

`lit` is defined has only using 24bit in the structure. In our case I switch to i32, it does take extra 8 bits, but solves the casting of usize to u16 of values greater then u16::MAX, which does happen in C++ version. I have built 2 toy projects, one in Rust and another in C++ (using the OpenEXR library) and use them to compare results.

```struct HufDec
{				// short code		long code
				//-------------------------------
    int		len:8;		// code length		0	 
    int		lit:24;		// lit			p size	 
    int	*	p;		// 0			lits	 
};
```

**Comment by johannesvollmer** [comment]
> But `int` is indeed the normal C++ int32 so `i32` is correct. 
Ok, cool.

> We did have a case where there we used u16 in the Code, and in C++ it's defined as an int but with only 24bits.

I made that change only recently, because it looked to me like all usages were casting it to an u16 sooner or later. It has been i32 before I changed it. I'm really sorry that I caused this bug! It was not smart of me to change that code when it was not working.

I'll try to make it rusty first before doing any optimizations in the future.

**Comment by johannesvollmer** [comment]
Also, I saw that you made comments about stack overflow that required the use of `vec![0; n]` instead of `[0; n]`. 

Has this actually been a problem on your machine? If this is the case, we should also use `vec!` in the wavelet.rs and piz.rs modules, which currently sometimes use `[0;n]`, resulting in possibly very large stack allocations.

**Comment by dgsantana** [comment]
> I made that change only recently, because it looked to me like all usages were casting it to an u16 sooner or later. It has been i32 before I changed it. I'm really sorry that I caused this bug! It was not smart of me to change that code when it was not working.
> 
Don't worry, I think we were already casting to u16, so the bug would appear anyway ;).

**Comment by dgsantana** [comment]
> 
> 
> Also, I saw that you made comments about stack overflow that lead to the use of `vec![0; n]` instead of `[0; n]`.
> 
> Has this actually been a problem on your machine? If this is the case, we should also use `vec!` in the wavelet.rs and piz.rs modules, which currently sometimes use `[0;n]`, resulting in possibly very large stack allocations.

I had stack overflows on Windows 10 x64 (64Gb RAM) on Rust 1.43.1-msvc. I don't recall if I had this problem on Linux (Pop!OS 20.04, same machine). I think it was on some methods that were creating 2 or 3 large stack arrays.

**Comment by johannesvollmer** [comment]
Ok, I guess if somebody notices a problem we can always change it at that point.

**Comment by johannesvollmer** [comment]
After three hours of debugging I remembered that NaN is not NaN and that the tests failed only because they contained values that were not the same, even though the algorithm actually worked and the values are actually the same. Thanks, NaN <3

**Comment by johannesvollmer** [comment]
All images pass now (I changed the tests to respect NaN (now they respect Nan more than NaN respects itself))

**Comment by johannesvollmer** [comment]
I have hot-fixed something in the huffman decoding, as it panicked while fuzzing - see [this commit](https://github.com/johannesvollmer/exrs/commit/ad28f9529590eae0491b73d09b9ed6d5c98dedcf)

https://github.com/johannesvollmer/exrs/blob/ad28f9529590eae0491b73d09b9ed6d5c98dedcf/src/compression/piz/huffman.rs#L253-L257

I have not yet investigated why this happens or what the difference to the reference implementation is, so this hot-fix might be incorrect in some way.

**Comment by johannesvollmer** [comment]
Ok so (as I am excited that it actually works) after testing and fuzzing a lot, I think this should be merged into master :)

**Comment by dgsantana** [comment]
Awesome. Already test the branch on our tool, and don't see any regressions, and the new compressions work and open fine on external tools. Thanks a lot for this crate, it removes a lot of the mess that is required to use OpenEXR in Rust. I will try to keep contributing ;).

**Comment by johannesvollmer** [comment]
I'm happy that this crate is of use. Thanks for you contribution, it is a pleasure to work with you. <3

**Comment by johannesvollmer** [comment]
I'm closing this issue, finally, feel free to open a new one for "rustifying" the compression code if there is something you want to talk about :)
